cpp/src/heap.cuh

```diff
@@ -0,0 +1,225 @@
+     *
+     *  Note, the heap will be represented like this - the
+     *  shape indicates the binary tree structure, the element
+     *  indicates the index of the array that is associated
+     *  with the element.  This diagram will help understand
+     *  the parent/child calculations defined below.
+     *
+     *                    0
+     *              1           2
+     *           3     4     5     6
+     *          7  8  9 10 11 12 13 14
+     *
+     *   So element 0 is the root of the tree, element 1 is the
+     *   left child of 0, element 2 is the right child of 0, etc.
+     */
+
+    namespace detail {
+      /**
+       * @brief Identify the parent index of the specified index.
+       *        NOTE: This function does no bounds checking, so
+       *        the parent of 0 is 0.
+       *
+       *   See the above documentation for a picture to describe
+       *   the tree.
+       *
+       * @param[in]  index - the current array index
+       * @return     the index of the parent of the current index
+       */
+      template <typename S>
+      inline S __host__ __device__ parent(S index) {
+	static_assert(std::is_integral<S>::value, "Index must be of an integral type");
+      
+	return ((index + 1) / 2) - 1;
+      }
+
+      /**
+       * @brief Identify the left child index of the specified index.
+       *        NOTE: This function does no bounds checking, so
+       *        the left child computed might be out of bounds.
+       *
+       *   See the above documentation for a picture to describe
+       *   the tree.
+       *
+       * @param[in]  index - the current array index
+       * @return     the index of the left child of the current index
+       */
+      template <typename S>
+      inline S __host__ __device__ left_child(S index) {
+	static_assert(std::is_integral<S>::value, "Index must be of an integral type");
+      
+	return ((index + 1) * 2 - 1);
+      }
+
+      /**
+       * @brief Identify the right child index of the specified index.
+       *        NOTE: This function does no bounds checking, so
+       *        the right child computed might be out of bounds.
+       *
+       *   See the above documentation for a picture to describe
+       *   the tree.
+       *
+       * @param[in]  index - the current array index
+       * @return     the index of the right child of the current index
+       */
+      template <typename S>
+      inline S __host__ __device__ right_child(S index) {
+	static_assert(std::is_integral<S>::value, "Index must be of an integral type");
+      
+	return (index + 1) * 2;
+      }
+    }
+      
+    /**
+     * @brief Reorder an existing array of elements into a heap
+     *
+     * @param[in, out]   array   - the existing array
+     * @param[in]        size    - the number of elements in the existing array
+     * @param[in]        compare - the comparison function to use
+     *
+     */
+    template <typename T, typename S, typename C>
```

Comment:
What are `typename T, typename S, typename C`? It would be great to document or rename with more explicit names.

Reply:
Suggestions on how to document this?  I don't see anything in the doxygen documentation about how to properly do that.  I guess I could rename them to ArrayT, SizeT and Compare if that seems more clear to you.

Reply:
I think just a regular comment would work or the renaming you suggested. It is just to help with readability for other developers if it needs to be modified or re-used elsewhere in the future.

Reply:
Just pushed this change.